pages/api/companies.ts

```diff
@@ -1,5 +1,130 @@
+import { PrismaClient } from '@prisma/client';
 import type { NextApiRequest, NextApiResponse } from 'next';
+import { CompanyData, CompanyListData, CompanyPostData } from '../../types/company';
+import { isEmpty, isValidUrl } from '../../utils/strings/validations';
+import { capitalizeEveryWord, removeProtocolAndWwwIfPresent } from '../../utils/strings/formatters';
+import { createJsonResponse, HttpMethod, HttpStatus, rejectHttpMethod } from '../../utils/http/httpHelpers';
+import { withAuth } from '../../utils/auth/jwtHelpers';
+import { ApiResponse, StatusMessageType } from '../../types/apiResponse';
+import { Nullable } from '../../types/utils';
+import { withPrismaErrorHandling } from '../../utils/prisma/prismaHelpers';
 
-export default function handler(req: NextApiRequest, res: NextApiResponse) {
-  res.status(200).end(`Request sent to ${req.url}\n`);
+enum MessageType {
+  COMPANY_ALREADY_EXISTS,
+  COMPANY_CREATED_SUCCESSFULLY,
+  EMPTY_NAME,
+  INVALID_COMPANY_URL,
+  MISSING_NAME,
+  MISSING_COMPANY_URL,
 }
+
+const prisma = new PrismaClient();
+
+const messages = Object.freeze({
+  [MessageType.COMPANY_ALREADY_EXISTS]: { type: StatusMessageType.SUCCESS, message: 'Company already exists.' },
+  [MessageType.COMPANY_CREATED_SUCCESSFULLY]: {
+    type: StatusMessageType.SUCCESS,
+    message: 'Company created successfully.',
+  },
+  [MessageType.EMPTY_NAME]: { type: StatusMessageType.ERROR, message: 'Company name is empty.' },
+  [MessageType.INVALID_COMPANY_URL]: { type: StatusMessageType.ERROR, message: 'Company url is invalid.' },
+  [MessageType.MISSING_NAME]: { type: StatusMessageType.ERROR, message: 'Company name is missing.' },
+  [MessageType.MISSING_COMPANY_URL]: { type: StatusMessageType.ERROR, message: 'Company url is missing.' },
+});
+
+function handler(req: NextApiRequest, res: NextApiResponse) {
+  const method = req.method;
+
+  switch (method) {
+    case HttpMethod.GET:
+      handleGet(req, res);
+      break;
+    case HttpMethod.POST:
+      handlePost(req, res);
+      break;
+    default:
+      rejectHttpMethod(res, method);
+  }
+}
+
+async function handleGet(req: NextApiRequest, res: NextApiResponse<ApiResponse<CompanyListData[]>>) {
+  if (req.body.searchKeyword === undefined || req.body.searchKeyword === null) {
+    res.status(HttpStatus.OK).json(createJsonResponse([]));
+    return;
+  }
+
+  const searchKeyword = req.body.searchKeyword.trim();
+
+  const companies: CompanyListData[] = await prisma.company.findMany({
+    where: {
+      isVerified: true,
+      name: {
+        contains: searchKeyword,
+        mode: 'insensitive',
+      },
+    },
```

Comment:
Do we want to support searching by company url as well? Smth like:

```suggestion
    where: {
      AND: [
        { isVerified: true },
       OR: [
         name: {
           contains: searchKeyword,
           mode: 'insensitive',
         },
         companyUrl: {
           contains: searchKeyword,
           mode: 'insensitive',
         }
       ] 
      ]
    },
```

Reply:
Same for the above comment, when will searching by url be useful? Users will only see the company name and want to find companies by name when making an application, they won't really see how the companyUrl is used in our app.